### DISCUSSION TITLE
What is the correct way to get a scoped DI object inside a singleton object?

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#7864] [Q&A] [answered]
Hello everyone

What is the correct way to get a scoped DI object inside singleton object?
I want to get `IContentDefinitionManager` object which is registered as scoped inside a singleton object `Migrations` and because of this I got an exception `InvalidOperationException: Cannot consume scoped service from singleton`

Therefore, is usage of IHttpContextAccessor  as the following correct? 
- Get IHttpContextAccessor as httpContextAccessorfrom constructor of singleton object    
- Then use httpContextAccessor as following or should I use other ways?

```
var serviceProvinder = httpContextAccessor.HttpContext.RequestServices;
var contentDefinitionManager = serviceProvinder.GetService(typeof(IContentDefinitionManager)) as ContentDefinitionManager;
```

Thank you.

**Comment by jtkech** [comment, marked as answer]
First try to organize your services and their lifetime differently, so that you don't need to inject a scoped service from a singleton. But in some case it might not be so easy, we have some in our repo, from memory `ISiteService` as it needs to be a singleton but also needs indirectly to use the YesSql `ISession` that is a scoped service.

In OrchardCore all is executed in a tenant context, and scoped services in a tenant scope (technically a `ShellScope`). During the same http request, some child scopes may be created, a tenant can create a scope on another tenant to resolve / use its services. So, more precisely you would need to use `ShellScope.Services` where `ShellScope` is an async state, as the value held by `IHttpContextAccessor`. That said, in OrchardCore we made the `HttpContext.RequestServices` tenant scope aware, so that you can still use it in the regular way to retrieve the services of the current shell scope.

So, if you can't find a way to not resolve a scoped service from a singleton, yes this is the right way. But take care that the methods of your singleton, that resolve and use a scoped service, are not called e.g. by another singleton outside a given shell scope

Don't cast to the `ContentDefinitionManager` implementation, just resolve the service through the interface.

    using Microsoft.Extensions.DependencyInjection;

    var services = _httpContextAccessor.HttpContext.RequestServices;
    var contentDefinitionManager = services.GetRequiredService<IContentDefinitionManager>();

**Reply by aaronamm** [reply]
@jtkech Thank you so much for the answer. It is very details and contains all information I need.
In addtion, thanks for letting me know there a generic version of GetRequiredService method.

**Reply by aaronamm** [reply]
@jtkech 
Very minor but I don't quite understand when you mentioned
"a tenant can create a scope on another tenant to resolve / use its services"

This makes me confuse because I think a tenant can create a scope on itself to resolve/use its service inside itself only

If you don't mind, could you please point me out the code in OC that a tenant create a scope on other tenant.

Thank you so much.🙏🙏

**Reply by jtkech** [reply]
@aaronamm 

> This makes me confuse because I think a tenant can create a scope on itself to resolve/use its service inside itself only

Most of the time this is the case

> If you don't mind, could you please point me out the code in OC that a tenant create a scope on other tenant.

E.g in `ModularBackgroundService.cs` where for each tenant we create a scope on it to resolve and then execute the registered `IBackgroundTask` of a given tenant. In the `OC.OpenId` module where different tenants can be the clients of another tenant acting as a server, see e.g. in `OpenIdValidationService.cs`. More globally, in the source code search for `GetScopeAsync` to find things like the following

                    var shellScope = await _shellHost.GetScopeAsync(shell.Settings);
                    await shellScope.UsingAsync(async scope =>
                    {
                        var someService = scope.ServiceProvider.GetServices<ISomeService>();
                        // some stuff
                    });

**Reply by aaronamm** [reply]
@jtkech Thank you so much for prompt reply and useful information.
